optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,277 @@
+from collections import OrderedDict
+from typing import Any
+from typing import Dict
+from typing import Optional
+
+import scipy
+
+import optuna
+from optuna import distributions
+from optuna import logging
+from optuna._experimental import experimental
+from optuna._transform import _SearchSpaceTransform
+from optuna.distributions import BaseDistribution
+from optuna.samplers import BaseSampler
+from optuna.study import Study
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+_logger = logging.get_logger(__name__)
+
+_SUGGESTED_STATES = (TrialState.COMPLETE, TrialState.PRUNED)
+
+_NUMERICAL_DISTRIBUTIONS = (
+    distributions.UniformDistribution,
+    distributions.LogUniformDistribution,
+    distributions.DiscreteUniformDistribution,
+    distributions.IntUniformDistribution,
+    distributions.IntLogUniformDistribution,
+)
+
+
+@experimental("2.x.0")  # TODO(kstoneriv3)
+class QMCSampler(BaseSampler):
+    """A Quasi Monte Carlo Sampler that generates low-discrepancy sequences.
+
+    Quasi Monte Carlo (QMC) sequences are designed to have low-discrepancies than
+    standard random seqeunces. They are known to perform better than the standard
+    randam sequences in hyperparameter optimization.
+
+    For further information about the use of QMC sequences for hyperparameter optimization,
+    please refer to the following paper:
+    - Bergstra, James, and Yoshua Bengio. Random search for hyper-parameter optimization.
+      Journal of machine learning research 13.2, 2012.
+      <https://jmlr.org/papers/v13/bergstra12a.html>`_
+
+    We use the QMC implementations in Scipy. For the details of the QMC algorithm,
+    see the Scipy API references on `scipy.stats.qmc
+    <https://scipy.github.io/devdocs/stats.qmc.html>`_.
+
+    .. note:
+        Please note that this sampler does not support CategoricalDistribution.
+        If your search space contains categorical parameters, it samples the catagorical
+        parameters at random without using QMC sequences.
```

Comment:
Ah, I should have documented that categorical variables are sampled by the `independent_sampler`, I guess. Does it well address your concern?

Reply:
Thanks, that sounds better actually. However, I also meant think you could omit the first sentence, but it's not a strong opinion.

Reply:
I changed my mind and will support the one-hot encoding of the categorical variable with some additional warnings. 
QMC might still be effective even if there exist categorical variables if they are not influential. Sobol sequence, for example, is effective when there are only a few variables that influence the target function. If we assume a situation where the categorical variable suggested does not influence the target function much, it makes good sense to use the Sobol sampler. When a categorical variable has a big influence on the target function, it makes more sense to run different Sobol samplers for each category.

Reply:
As for the naive support of the relative sampling of the categorical distribution, I will take some benchmarks to see if it works.

Reply:
# Benchmark Result Report

## Overall Results


| Solver                                                                                                   | Borda | Firsts |
|:---------------------------------------------------------------------------------------------------------|------:|-------:|
| [MultivariateTPE](#id-7e8f2dbfac3414616c7ef688df24c2a45c5c9dfeffb238a120dced1793a0932a)                  |    16 |      4 |
| [RandomSampler](#id-d20c9e8b094127d2961c2080859699a76f1d3439ba861d1d412dafc2f1d58fa9)                    |     0 |      0 |
| [SobolSampler](#id-b75c63eb93438532fdcd7b2c33127cdf783361b7b39192d4a29a0ee357b01abe)                     |     8 |      0 |
| [SobolSampler_without_Categorical](#id-68ab39b32632da5253e3edba66e19e0a7e0ec52a236691e6df7bf6cb6802faf1) |     8 |      0 |
| [TPE](#id-b68e3768ef212bbb3073d6ab200be8d6e4990c377a0849aba1ea96856f4b3c58)                              |     4 |      0 |

## Individual Results


### (1) Problem: [HPO-Bench-Slice](#id-52519b66083304cdb9ab43c80265c15100adcad0e23076e20b9462fa597d8b29)

| Ranking | Solver                                                                                                                                                                                   |     Best (avg +- sd) | AUC (avg +- sd) | Elapsed (avg +- sd) |
|--------:|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------:|----------------:|--------------------:|
|       1 | [MultivariateTPE](#id-7e8f2dbfac3414616c7ef688df24c2a45c5c9dfeffb238a120dced1793a0932a) ([study](#id-60c3b8dd12e3740723a162facf2e3a78d44f8a6c6ea8dbc5e6b8ec9ca757b944))                  | 0.000259 +- 0.000046 |  0.027 +- 0.007 |    75.754 +- 43.171 |
|       2 | [SobolSampler_without_Categorical](#id-68ab39b32632da5253e3edba66e19e0a7e0ec52a236691e6df7bf6cb6802faf1) ([study](#id-567a34d66227ec8356be2e8510f26116b9b24a1ab6d55f1bbab88bb43fc4c9d6)) | 0.000288 +- 0.000041 |  0.032 +- 0.005 |     36.079 +- 8.160 |
|       2 | [SobolSampler](#id-b75c63eb93438532fdcd7b2c33127cdf783361b7b39192d4a29a0ee357b01abe) ([study](#id-5838a9cf4180463143b416ab37de425eab29eb967475bff84ea2e89a8f076d26))                     | 0.000274 +- 0.000031 |  0.030 +- 0.004 |     32.926 +- 6.186 |
|       4 | [TPE](#id-b68e3768ef212bbb3073d6ab200be8d6e4990c377a0849aba1ea96856f4b3c58) ([study](#id-ea1c7771bd5dee5a2179a68290a484de6a619aa3a63b8b7040127c35d0e35c3c))                              | 0.000355 +- 0.000132 |  0.037 +- 0.014 |     30.661 +- 7.882 |
|       5 | [RandomSampler](#id-d20c9e8b094127d2961c2080859699a76f1d3439ba861d1d412dafc2f1d58fa9) ([study](#id-8b03b4d8c956d0d7c5b41dbca9760577a368ec0cf7b008b3ae4e48afa9d8474b))                    | 0.000456 +- 0.000170 |  0.065 +- 0.031 |      0.120 +- 0.020 |

### (2) Problem: [HPO-Bench-Naval](#id-6b72e8e45774af501b8d501125a908f6d2ed4220028ee71fbf5bded76d662bc4)

| Ranking | Solver                                                                                                                                                                                   |     Best (avg +- sd) | AUC (avg +- sd) | Elapsed (avg +- sd) |
|--------:|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------:|----------------:|--------------------:|
|       1 | [MultivariateTPE](#id-7e8f2dbfac3414616c7ef688df24c2a45c5c9dfeffb238a120dced1793a0932a) ([study](#id-66106fabcb393d5e2ad68e3524fbb180204d0b2d8856ca06f858fd472d545b24))                  | 0.000039 +- 0.000026 |  0.004 +- 0.006 |    49.076 +- 30.788 |
|       2 | [SobolSampler_without_Categorical](#id-68ab39b32632da5253e3edba66e19e0a7e0ec52a236691e6df7bf6cb6802faf1) ([study](#id-0f62dec4615b8597cd56113782334c6d89fd5878e9c11f8e65acac3d98c79df3)) | 0.000041 +- 0.000008 |  0.004 +- 0.001 |     36.627 +- 7.105 |
|       2 | [SobolSampler](#id-b75c63eb93438532fdcd7b2c33127cdf783361b7b39192d4a29a0ee357b01abe) ([study](#id-4c8b9d17a9a7f43934e1785b577cfe8ef66aa7d45cf53bb85a60759b1c6ca758))                     | 0.000039 +- 0.000007 |  0.004 +- 0.001 |     34.730 +- 5.944 |
|       4 | [TPE](#id-b68e3768ef212bbb3073d6ab200be8d6e4990c377a0849aba1ea96856f4b3c58) ([study](#id-a9c6502c1468e982f9da2e9bf1d178c3ff4eef77d773594c306ffde61247ff7e))                              | 0.000115 +- 0.000230 |  0.011 +- 0.020 |     27.741 +- 8.005 |
|       5 | [RandomSampler](#id-d20c9e8b094127d2961c2080859699a76f1d3439ba861d1d412dafc2f1d58fa9) ([study](#id-e1bdf815e310bf57647673998019c11e237e9bf6cf18781896f8ecb572addc14))                    | 0.000149 +- 0.000176 |  0.033 +- 0.036 |      0.117 +- 0.019 |

### (3) Problem: [HPO-Bench-Protein](#id-a777567ecdcc72a37a4ae27a1f5694120ab8b86c89ac276b8d4a82586d4d5a66)

| Ranking | Solver                                                                                                                                                                                   |     Best (avg +- sd) | AUC (avg +- sd) | Elapsed (avg +- sd) |
|--------:|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------:|----------------:|--------------------:|
|       1 | [MultivariateTPE](#id-7e8f2dbfac3414616c7ef688df24c2a45c5c9dfeffb238a120dced1793a0932a) ([study](#id-8b77a4a0eee6d3010ca8bcf40632430260dfef87a49edb0b85f7be18a987ed92))                  | 0.222753 +- 0.002474 | 18.199 +- 0.237 |    40.335 +- 19.875 |
|       2 | [SobolSampler_without_Categorical](#id-68ab39b32632da5253e3edba66e19e0a7e0ec52a236691e6df7bf6cb6802faf1) ([study](#id-347a70325b8fe065f3ca44b505472d5bf35c6c858cbc994e07f3bff0dc9f5aae)) | 0.226416 +- 0.003629 | 18.713 +- 0.390 |     36.221 +- 7.424 |
|       2 | [SobolSampler](#id-b75c63eb93438532fdcd7b2c33127cdf783361b7b39192d4a29a0ee357b01abe) ([study](#id-c22934e092a9241cbf707a3896ec30e662d715ee8480e49cfa9eeab904f03875))                     | 0.225859 +- 0.002818 | 18.616 +- 0.290 |     33.630 +- 5.403 |
|       4 | [TPE](#id-b68e3768ef212bbb3073d6ab200be8d6e4990c377a0849aba1ea96856f4b3c58) ([study](#id-e8d23b7ecc65daa011a39c6c5ffd73c90bfb106ff21df7779ee0b101e656b4a4))                              | 0.231648 +- 0.007876 | 19.006 +- 0.769 |     25.673 +- 5.940 |
|       5 | [RandomSampler](#id-d20c9e8b094127d2961c2080859699a76f1d3439ba861d1d412dafc2f1d58fa9) ([study](#id-e787cd65ef00c2c7dce3f29ed19d8b3a06dd1ea6963e2471cbbd7888344d5263))                    | 0.254443 +- 0.015337 | 21.444 +- 1.259 |      0.116 +- 0.017 |

### (4) Problem: [HPO-Bench-Parkinson](#id-fb768edf855fcb1277b511c212e74ae5ece3abd16a2029aeacd041e5cf3eb7ec)

| Ranking | Solver                                                                                                                                                                                   |     Best (avg +- sd) | AUC (avg +- sd) | Elapsed (avg +- sd) |
|--------:|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------:|----------------:|--------------------:|
|       1 | [MultivariateTPE](#id-7e8f2dbfac3414616c7ef688df24c2a45c5c9dfeffb238a120dced1793a0932a) ([study](#id-b84067ca23735b93dfa6dea35c7de182888fd9ea7197a1e36ee5b6348eacefcb))                  | 0.007736 +- 0.001696 |  0.861 +- 0.172 |    66.651 +- 38.781 |
|       2 | [SobolSampler_without_Categorical](#id-68ab39b32632da5253e3edba66e19e0a7e0ec52a236691e6df7bf6cb6802faf1) ([study](#id-8d945a02b3d94d24b25160d3eb3f37b6501c19d2de7be635e5344c5df0c64f09)) | 0.008482 +- 0.001555 |  1.002 +- 0.167 |     36.416 +- 6.063 |
|       2 | [SobolSampler](#id-b75c63eb93438532fdcd7b2c33127cdf783361b7b39192d4a29a0ee357b01abe) ([study](#id-0a1dafdcf8a48777c63ab0177aaef2ed82fb059854609794abf9ca7a03e179f2))                     | 0.008768 +- 0.001601 |  0.989 +- 0.149 |     34.046 +- 5.428 |
|       4 | [TPE](#id-b68e3768ef212bbb3073d6ab200be8d6e4990c377a0849aba1ea96856f4b3c58) ([study](#id-7976d79b907710c9d666723b4161b9ba440d224ae454d348181cba6755563cbd))                              | 0.011679 +- 0.003545 |  1.227 +- 0.367 |     27.503 +- 7.301 |
|       5 | [RandomSampler](#id-d20c9e8b094127d2961c2080859699a76f1d3439ba861d1d412dafc2f1d58fa9) ([study](#id-04c7c22b1eea1e941609fcba208017c97718bfffdbf95d18c1469b9f2dc2802f))                    | 0.014856 +- 0.004492 |  1.820 +- 0.603 |      0.117 +- 0.020 |

## Plots

![hpo-bench-naval-6b72e8e45774af501b8d501125a908f6d2ed4220028ee71fbf5bded76d662bc4](https://user-images.githubusercontent.com/30857855/110085956-09ea9d80-7d92-11eb-855f-03e575f553ed.png)
![hpo-bench-parkinson-fb768edf855fcb1277b511c212e74ae5ece3abd16a2029aeacd041e5cf3eb7ec](https://user-images.githubusercontent.com/30857855/110085959-0a833400-7d92-11eb-8d5d-f8aff8454d67.png)
![hpo-bench-protein-a777567ecdcc72a37a4ae27a1f5694120ab8b86c89ac276b8d4a82586d4d5a66](https://user-images.githubusercontent.com/30857855/110085961-0a833400-7d92-11eb-9c1b-757979d09e45.png)
![hpo-bench-slice-52519b66083304cdb9ab43c80265c15100adcad0e23076e20b9462fa597d8b29](https://user-images.githubusercontent.com/30857855/110085962-0b1bca80-7d92-11eb-8a3b-3efa3ade1b7d.png)

Reply:
For me, there seems to be no difference in sampler w/o naive support for categorical distribution.... Still, a sampler with naive support for categorical distribution tends to be very slightly better, though there is no statistically significant difference.

Reply:
Thanks for the benchmarks and the interesting results. HPOBench only contains discrete and categorical parameters so even more so. Personally, I'd given this prefer supporting categorical natively since that should keep the code simpler.

Reply:
Sobol I (1993) Sensitivity estimates for non linear mathematical models. Mathematical Modelling and Computational Experiments 1:407–414

Reply:
There is a discussion on whether it is reasonable to use Sobol sequence for Discrete variables... and the above paper was mentioned there.  [https://stats.stackexchange.com/questions/190843/sobol-indices-for-discrete-variables](https://stats.stackexchange.com/questions/190843/sobol-indices-for-discrete-variables)

Reply:
\* My implementation of a categorical variable is not one-hot. I took a kind of the inverse of the cumulative distribution function. (e.g. when sampling from [0,1,2,3], `param_value = int(np.ceil(4*sobol_sample))`)

Reply:
I did experiments with Halton and Latin Hypercube and the result was the same: there is no difference in performance...